modeshape-jcr/src/main/java/org/modeshape/jcr/txn/NoClientTransactions.java

```diff
@@ -50,14 +50,15 @@ public Transaction currentTransaction() {
 
     @Override
     public synchronized Transaction begin() throws NotSupportedException, SystemException {
-        if (ACTIVE_TRANSACTION.get() == null) {
+        NestableThreadLocalTransaction tx = ACTIVE_TRANSACTION.get();
+        if (tx == null) {
             // Start a transaction ...
             txnMgr.begin();
             if (logger.isTraceEnabled()) {
                 logger.trace("Begin transaction {0}", currentTransactionId());
             }
-            ACTIVE_TRANSACTION.set(new NestableThreadLocalTransaction(txnMgr, ACTIVE_TRANSACTION));
+            tx = new NestableThreadLocalTransaction(txnMgr, ACTIVE_TRANSACTION);
         }
-        return ACTIVE_TRANSACTION.get().begin();
+        return tx.begin();
```

Comment:
The cleanup code I mentioned was in `NestedThreadLocalTransaction` [here](https://github.com/ModeShape/modeshape/pull/1450/files#diff-971a87033e57696cf8bfee58d9a4ce74R417), so that definitely applies.